File: ch03/boolean.py
# Example Boolean functions

def f(x,y):
    return not x and (x or not y)

print("x y f")
print("-----")
for x in [0,1]:
    for y in [0,1]:
        print("%d %d %d" % (x,y,f(x,y)))
print()

def fc(x,y):
    return x or not x and y

print("x y fc")
print("------")
for x in [0,1]:
    for y in [0,1]:
        print("%d %d %d" % (x,y,fc(x,y)))
print()

def g(a,b,c):
    return (a or b and not c) and (a and b or not b and c)

print("a b c g")
print("-------")
for a in [0,1]:
    for b in [0,1]:
        for c in [0,1]:
            print("%d %d %d %d" % (a,b,c,g(a,b,c)))
print()

def gc(a,b,c):
    return not a and (not b or c) or (not a or not b) and (b or not c)

print("a b c gc")
print("--------")
for a in [0,1]:
    for b in [0,1]:
        for c in [0,1]:
            print("%d %d %d %d" % (a,b,c,gc(a,b,c)))
print()

def gd(a,b,c):
    return (a and (b or not c)) or ((a or b) and (not b or c))

print("a b c gd")
print("--------")
for a in [0,1]:
    for b in [0,1]:
        for c in [0,1]:
            print("%d %d %d %d" % (a,b,c,gd(a,b,c)))
print()

def e(a,b,c):
    return (a and not b and c) or (not a and b and not c) or (a and b and c)

print("a b c e")
print("-------")
for a in [0,1]:
    for b in [0,1]:
        for c in [0,1]:
            print("%d %d %d %d" % (a,b,c,e(a,b,c)))
print()

def ed(a,b,c):
    return (a or not b or c) and (not a or b or not c) and (a or b or c)

print("a b c ed")
print("--------")
for a in [0,1]:
    for b in [0,1]:
        for c in [0,1]:
            print("%d %d %d %d" % (a,b,c,ed(a,b,c)))
print()

def ec(a,b,c):
    return (not a or b or not c) and (a or not b or c) and (not a or not b or not c)

print("a b c ec")
print("--------")
for a in [0,1]:
    for b in [0,1]:
        for c in [0,1]:
            print("%d %d %d %d" % (a,b,c,ec(a,b,c)))

File: ch03/boolean.c
// Example Boolean functions
#include <stdio.h>
#include <inttypes.h>

uint8_t f(uint8_t x, uint8_t y) {
    return !x && (x || !y);
}

uint8_t fd(uint8_t x, uint8_t y) {
    return !x || (x && !y);
}

uint8_t fc(uint8_t x, uint8_t y) {
    return x || (!x && y);
}

uint8_t g(uint8_t a, uint8_t b, uint8_t c) {
    return (a || b&&!c) && (a&&b || !b&&c);
}

int main() {
    uint8_t x,y,a,b,c;
    
    printf("x y f\n-----\n");
    for(x=0; x<2; x++)
        for(y=0; y<2; y++)
            printf("%d %d %d\n", x, y, f(x,y));
    printf("\n");

    printf("x y fd\n-----\n");
    for(x=0; x<2; x++)
        for(y=0; y<2; y++)
            printf("%d %d %d\n", x, y, fd(x,y));
    printf("\n");

    printf("x y fc\n-----\n");
    for(x=0; x<2; x++)
        for(y=0; y<2; y++)
            printf("%d %d %d\n", x, y, fc(x,y));
    printf("\n");

    printf("a b c g\n-------\n");
    for(a=0; a<2; a++)
        for(b=0; b<2; b++)
            for(c=0; c<2; c++)
                printf("%d %d %d %d\n", a,b,c, g(a,b,c));
    printf("\n");

    return 0;
}

